src/header-parser.ts

```diff
@@ -0,0 +1,336 @@
+      break;
+    }
+    case 'implementation-defined abstract operation': {
+      para.innerHTML += `The implementation-defined abstract operation ${name} takes ${formattedParams}.`;
+      break;
+    }
+    case 'internal method':
+    case 'concrete method': {
+      if (_for == null) {
+        spec.warn({
+          type: 'contents',
+          ruleId: 'header-format',
+          message: `expected ${type} to have a "for"`,
+          node: dl,
+          nodeRelativeLine: 1,
+          nodeRelativeColumn: 1,
+        });
+        _for = spec.doc.createElement('div');
+      }
+      para.append(
+        `The ${name} ${type} of `,
+        // @ts-ignore childNodes is iterable
+        ..._for.childNodes
+      );
+      para.innerHTML += ` takes ${formattedParams}.`;
+      break;
+    }
+    default: {
+      if (type === 'unknown') {
+        spec.warn({
+          type: 'node',
+          ruleId: 'header-type',
+          message: `clauses with structured headers should have a type`,
+          node: clause,
+        });
+      } else {
+        spec.warn({
+          type: 'attr',
+          ruleId: 'header-type',
+          message: `unknown clause type ${type}`,
+          node: clause,
+          attr: 'type',
+        });
+      }
+    }
+  }
+  if (description != null) {
+    // @ts-ignore childNodes is iterable
+    const isJustElements = [...(description.childNodes as Iterable<Node>)].every(
+      n => n.nodeType === 1 || (n.nodeType === 3 && n.textContent?.trim() === '')
+    );
+    if (isJustElements) {
+      // @ts-ignore childNodes is iterable
+      paras.push(...description.childNodes);
+    } else {
+      // @ts-ignore childNodes is iterable
+      para.append(' ', ...description.childNodes);
+    }
+  }
+  let next = dl.nextElementSibling;
+  while (next != null && next.tagName === 'EMU-NOTE') {
+    next = next.nextElementSibling;
+  }
+  if (next?.tagName == 'EMU-ALG' && !next.hasAttribute('replaces-step')) {
+    if (paras.length > 1 || next !== dl.nextElementSibling) {
+      const whitespace = next.previousSibling;
+      const after = spec.doc.createElement('p');
+      after.append('It performs the following steps when called:');
+      next.parentElement!.insertBefore(after, next);
+
+      // fix up the whitespace in the generated HTML
+      if (whitespace?.nodeType === 3 /* TEXT_NODE */ && /^\s+$/.test(whitespace.nodeValue!)) {
+        next.parentElement!.insertBefore(whitespace.cloneNode(), next);
+      }
+    } else {
+      para.append(' It performs the following steps when called:');
+    }
+  }
+  return paras;
+}
+
+function eat(text: string, regex: RegExp) {
+  const match = text.match(regex);
+  if (match == null) {
+    return { success: false, match, text };
+  }
+  return { success: true, match, text: text.substring(match[0].length) };
+}
+
+function formatEnglishList(list: Array<string>) {
+  if (list.length === 1) {
```

Comment:
Maybe throw for `list.length === 0`.

Reply:
Done. Would be nice if the type system could handle that, but I appreciate the difficulty involved.